### ISSUE TITLE
XNNPack delegate support for quantized models, no latency improvement

### ISSUE THREAD

**Issue by Soleimani64** [tensorflow/tensorflow#55525]
I built tensorflow with  --define tflite_with_xnnpack=true --define xnn_enable_qs8=true to have acceleration for quantized models. But I don't gain any improvement in the latency of the quantized models compared to when I build tensorflow with --define tflite_with_xnnpack=true alone. I tested both in windows and linux. I am testing on desktop with intel CPU.  Is there anything I am doing wrong? I tested with trained models too. I used the following code to quantize the model.:

```
import tensorflow as tf
import numpy as np
from tensorflow.keras.layers import *

from os import listdir
from os.path import isfile, join

def representative_dataset():
    folder = '.\\test'
    onlyfiles = [f for f in listdir(folder) if isfile(join(folder, f))]  #you can use random arrays I guess
    for file in onlyfiles:
        data = np.load(join(folder, file))
        data = np.expand_dims(data, axis=0)
        yield [tf.dtypes.cast(data, tf.float32)]

def SOC(input_tensor, classNumer=1, epsilonBN=1e-3):
    a = Input(shape=input_tensor)
    x = Conv2D(16, (3, 3), padding='same', use_bias=False, name="Conv_1")(a)
    x = BatchNormalization(epsilon=epsilonBN, name="BN_1")(x)
    x=Activation('relu')(x)
    x = MaxPooling2D(pool_size=(2, 2), name="MaxPool_1")(x)

    x = Conv2D(32, (3, 3), padding='same', use_bias=False, name="Conv_2")(x)
    x = BatchNormalization(epsilon=epsilonBN, name="BN_2")(x)
    x=Activation('relu')(x)
    x = MaxPooling2D(pool_size=(2, 2), name="MaxPool_2")(x)

    x = Conv2D(64, (3, 3), padding='same', use_bias=False, name="Conv_3")(x)
    x = BatchNormalization(epsilon=epsilonBN, name="BN_3")(x)
    x=Activation('relu')(x) 
    x = MaxPooling2D(pool_size=(2, 2), name="MaxPool_3")(x)

    x = Conv2D(128, (3, 3), padding='same', use_bias=False, name="Conv_4")(x)
    x = BatchNormalization(epsilon=epsilonBN, name="BN_4")(x)
    x = Activation('relu')(x)
    x = MaxPooling2D(pool_size=(2, 2), name="MaxPool_4")(x)

    x = Flatten()(x)
    x = Dense(classNumer, kernel_initializer='uniform', name="Dense")(x)
    model = tf.keras.Model(inputs=a, outputs=x)
    return model

new_model = SOC((32,32,5))
converter = tf.lite.TFLiteConverter.from_keras_model(new_model)
converter.optimizations = [tf.lite.Optimize.DEFAULT]
converter.representative_dataset = representative_dataset
tflite_quant_model = converter.convert()

with open('qmodel.tflite', 'wb') as f:
    f.write(tflite_quant_model)
````

**System information**
- OS: Windows 10 Pro
- Desktop, Intel(R) Core(TM) i7-9700K CPU 
- Tensorflow 2.9 built from source (tested with tensoflow 2.8 built from source on linux too)
- python 3.10
- Bazel 5.1.0
- Did not include gpu support on the tensorflow build in windows

**Comment by mohantym** [comment]
Hi @Soleimani64 ! According to this [document](https://blog.tensorflow.org/2021/09/faster-quantized-inference-with-xnnpack.html#:~:text=server%2Dclass%20processors.-,How%20can%20you%20use%20it%3F,-Quantized%20XNNPACK%20inference), Limited support for operators with asymmetric quantization is available via the --define xnn_enable_qu8=true Bazel option. So **"-define tflite_with_xnnpack=true --define xnn_enable_qs8=true"** correct command to build with bazel.Thanks!

**Comment by Soleimani64** [comment]
Hi Mohantym,
Thanks for the reply. but I used  "**-define tflite_with_xnnpack=true --define xnn_enable_qs8=true**" to build tensorflow. I did **not** use --define xnn_enable_qu8=true

**Comment by Soleimani64** [comment]
@mohantym
To make it clear, the problem here is that I was expecting to see lower latency for quantized models when I use "--define xnn_enable_qs8=true" option. But the latency doesn't change at all. I was wondering if you can give me an example when "--define xnn_enable_qs8=true" actually reduces the latency of a quantized model.
 I assumed that with this option we can use fixed-point operations on a desktop cpu. Is that correct?

**Comment by mohantym** [comment]
Sorry @Soleimani64 ! According to this [document](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/delegates/xnnpack/README.md), **BatchNormalization** is not supported operator yet in XNNPack yet. I think you have to stay with **-define tflite_with_xnnpack=true** for now to use the default XNNPack engine . Thanks!

**Comment by Soleimani64** [comment]
Thanks @mohantym 
I tried without BatchNormalization, the latency became a little bit smaller with --define xnn_enable_qs8=true, but the difference is very small. I will try with bigger models.
How about my other question: does tensorflow do 8-bit operations when we activate --define xnn_enable_qs8=true, or it still uses the floating point?

**Comment by mohantym** [comment]
Hi @sachinprasadhs ! Could you please look at this issue?

**Comment by Maratyszcza** [comment]
@Soleimani64 use `--define tflite_with_xnnpack_qs8=true` and/or `--define tflite_with_xnnpack_qu8=true` Bazel flags.
See [XNNPack delegate documentation](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/lite/delegates/xnnpack#quantized-operators) for details.

**Comment by Soleimani64** [comment]
@Maratyszcza I did that, it doesn't improve the latency.

**Comment by Maratyszcza** [comment]
Do you use the latest version of TensorFlow Lite?
Do you see "Created TensorFlow Lite XNNPACK delegate for CPU." message when you run the model?

**Comment by Soleimani64** [comment]
@Maratyszcza , yes to both questions. 
I can see improvement in the latency of both tf32 and int8 models when comparing XNN pack with regular tensorflow. but I don't see any difference in the latency of the int8 models if I use --define tflite_with_xnnpack_qs8=true and --define tflite_with_xnnpack_qu8=true compared to when I don't use them.

**Comment by huzq85** [comment]
> Hi @Soleimani64 ! According to this [document](https://blog.tensorflow.org/2021/09/faster-quantized-inference-with-xnnpack.html#:~:text=server%2Dclass%20processors.-,How%20can%20you%20use%20it%3F,-Quantized%20XNNPACK%20inference), Limited support for operators with asymmetric quantization is available via the --define xnn_enable_qu8=true Bazel option. So **"-define tflite_with_xnnpack=true --define xnn_enable_qs8=true"** is correct command to build with bazel here .
> 
> Attaching relevant [thread](https://www.tensorflow.org/model_optimization/guide/pruning/pruning_for_on_device_inference) for reference. Thanks!

hi, I just checked the 

> Do you use the latest version of TensorFlow Lite? Do you see "Created TensorFlow Lite XNNPACK delegate for CPU." message when you run the model?

Actually I encountered the same issue. I can see "Created Tensorflow Lite XNNPACK delegate for CPU" but the thing is that when running on the floating model, there are multiple lines of "Created Tensorflow Lite XNNPACK ...", for the int8 model, there is only one such line printed.

**Comment by huzq85** [comment]
> Do you use the latest version of TensorFlow Lite? Do you see "Created TensorFlow Lite XNNPACK delegate for CPU." message when you run the model?

Actually I encountered the same issue. I can see "Created Tensorflow Lite XNNPACK delegate for CPU" but the thing is that when running on the floating model, there are multiple lines of "Created Tensorflow Lite XNNPACK ...", for the int8 model, there is only one such line printed.

And, the quantized (int8) model obtains much more latency than the floating point model.

**Comment by sachinprasadhs** [comment]
Hi could you please compile the package with debug mode with something like below to get more verbose logs so that we can investigate further on the issue.
`bazel build -c dbg --copt=-g  --jobs=8 //tensorflow/tools/pip_package:build_pip_package`
